### PR TITLE
Avoid another race condition of open mode

### PR DESCRIPTION
**Pull request by nobu**
## What was the end-user or developer problem that led to this PR?

https://github.com/rubygems/rubygems/pull/7806#discussion_r1698091858
> This seems a race condition. Why not `IO::RDWR | IO::APPEND | IO::CREAT` simply?

## What is your fix for the problem, implemented in this PR?

Instead, just open in CREATE and APPEND mode.
Also, move the workaround for old Solaris as fallback to retry.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Review by deivid-rodriguez — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
We used to special case Solaris before https://github.com/rubygems/rubygems/commit/27b682c81226838b1254ac5843a3f5b1cb20f076, but since we're no longer doing anything special there I think, and we've got no issues reported? Can you explain more about the changes related to Solaris?

**Comment by nobu** [comment]
> We used to special case Solaris before [27b682c](https://github.com/rubygems/rubygems/commit/27b682c81226838b1254ac5843a3f5b1cb20f076), but since that commit we're no longer doing anything special there I think, and we've got no issues reported? Can you explain more about the changes related to Solaris?

That might have been fix in newer versions of Solaris.
Or, I guess no one runs Ruby on Solaris nowadays ;-)

Anyway, removed that case.

File: lib/rubygems.rb [hunk 1]
```diff
@@ -798,24 +798,20 @@ def self.open_file(path, flags, &block)
     File.open(path, flags, &block)
   end
 
+  MODE_TO_FLOCK = IO::RDONLY | IO::APPEND | IO::CREAT # :nodoc:
```
Comment (deivid-rodriguez) [inline lib/rubygems.rb:801]: Now I think it'd be more readable to remove this constant and pass the mode directly to the `File.open` call, given it's the only place using it?

Comment (deivid-rodriguez) [reply lib/rubygems.rb:801]: @nobu Let me know what you think. I feel the code is easier to follow if the mode used is right in there, but if you have a very strong opinion about using a constant, I don't mind. I'd make it private if we end up using a constant.

Comment (nobu) [reply lib/rubygems.rb:801]: It is for #7939.
`IO::SHARE_DELETE` is necessary on Windows, but it is not defined in TruffleRuby.

**Review by deivid-rodriguez — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
I forgot that we should backport this changes to https://github.com/rubygems/rubygems/blob/7ee2122f38163e4741d9aaf862370a818d14f58f/bundler/lib/bundler/rubygems_ext.rb#L33-L52. We can do it after #7939.

**Comment by eregon** [comment]
`flock(2)` is not that portable and e.g. old Solaris had significant bugs with it.
I wouldn't expect RubyGems to rely on such a low-level under-specified API.
What about using `open(O_CREAT|O_EXCL)` instead, shouldn't that be enough, at least for lock files? I would think that's far better supported and don't need all the workarounds and fallbacks here.

**Comment by deivid-rodriguez** [comment]
That makes sense, can you create a patch?

**Pull request merged**